source/common/upstream/upstream_impl.cc

```diff
@@ -1098,25 +1097,30 @@ LegacyLbPolicyConfigHelper::getTypedLbConfigFromLegacyProtoWithoutSubset(
                     ClusterProto::LbPolicy_Name(cluster.lb_policy())));
   }
 
-  return Result{lb_factory, lb_factory->loadConfig(factory_context, cluster)};
+  auto lb_config_or_error = lb_factory->loadLegacy(factory_context, cluster);
+  RETURN_IF_NOT_OK_REF(lb_config_or_error.status());
+  return Result{lb_factory, std::move(lb_config_or_error.value())};
 }
 
 absl::StatusOr<LegacyLbPolicyConfigHelper::Result>
 LegacyLbPolicyConfigHelper::getTypedLbConfigFromLegacyProto(
     Server::Configuration::ServerFactoryContext& factory_context, const ClusterProto& cluster) {
-  // Handle the lb subset config case first.
-  // Note it is possible to have a lb_subset_config without actually having any subset selectors.
-  // In this case the subset load balancer should not be used.
-  if (cluster.has_lb_subset_config() && !cluster.lb_subset_config().subset_selectors().empty()) {
-    auto* lb_factory = Config::Utility::getFactoryByName<TypedLoadBalancerFactory>(
-        "envoy.load_balancing_policies.subset");
-    if (lb_factory != nullptr) {
-      return Result{lb_factory, lb_factory->loadConfig(factory_context, cluster)};
-    }
+  if (!cluster.has_lb_subset_config() ||
```

Comment:
ooc why the order?  just makes the diffs harder to read

Reply:
because I hope the simpler code branch could be placed in quick return, and more complex code branch be handled in following function body.
I can revert it if you think it necessary.